### ISSUE TITLE
Missing stack trace when exception occurs during OneTimeSetUp

### ISSUE THREAD

**Issue by OsirisTerje** [nunit/nunit#2466]
@ericnewton76 commented on [Sat Dec 17 2016](https://github.com/nunit/nunit3-vs-adapter/issues/284)

![image](https://cloud.githubusercontent.com/assets/2211788/21288540/103270f6-c454-11e6-89ff-5ef94bd68ebb.png)

Missing the stack trace for the error if it bubbles up from a OneTimeSetUp method.

---

@ericnewton76 commented on [Sat Dec 17 2016](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-267782375)

I have forked vs-adapter and I'm looking through the code, and this type of scenario isn't tested.

https://github.com/ericnewton76/nunit3-vs-adapter/commit/09901cd4d26afbb4aa6dc02f28ed79c5abb9556c
I added a class to mock-assembly project (at the end of MockAssembly.cs) but i'm not sure how to run the tests against this new mock assembly.   

Additionally I wonder if nunit itself has the stack trace for the exception during OneTimeSetUp and if the vs-adapter is displaying all that it has.

---

@ericnewton76 commented on [Sat Dec 17 2016](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-267783470)

Ok, attempted to run build script, because I see it includes test running, but MSBuild giving me some kind of VS2017 sxs installation problem, so dont install VS2017 yet.  The VS developers are lamenting the fact that VS2015 cant build VS2017 at the moment.  I'm guessing this is related to my problem. with getting msbuild to build successfully.  https://github.com/github/VisualStudio/issues/655#issuecomment-261895992

---

@OsirisTerje commented on [Sun Dec 18 2016](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-267813809)

I can repro this, and I see we don't add the OneTimeSetup crash information to the output.  The information comes from the TestFixture node, and the site info is "Child". 
 @CharliePoole @rprouse : I see the Setup and Teardown have its own site names, but OneTimeSetup and OneTimeTearDown are not listed there.  And I can't see any of these site names in the output result file.  I only see Child and Parent.   

---

@CharliePoole commented on [Sun Dec 18 2016](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-267816835)

One time setup is the parent setup. It's difficult to know what to do here since VS has no place to report suite results but only test cases. We could report the suite result on each test case but that's potentially a lot of duplication. I compromised by reporting only the message but the stack trace could be reported as well.

---

@ericnewton76 commented on [Sun Dec 18 2016](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-267831160)

How does nunit handle an exception in onetimesetup?  Does it basically mark the whole test fixture as invalid, or does it keep retrying onetimesetup for each test in the fixture (Which to me would makes most sense)

---

@rprouse commented on [Sun Dec 18 2016](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-267833767)

It marks the whole fixture as failed, it doesn't retry. We assume that your setup code is deterministic and will fail consistently.

---

@OsirisTerje commented on [Sun Dec 25 2016](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-269140754)

Is the output I see in the console output (testresult.xml) identical to what the adapter gets as a response?

---

@CharliePoole commented on [Mon Dec 26 2016](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-269141326)

Yes, the adapter gets the entire XML result in pieces as each test completes. (It could also get the completed result but it doesn't need it since the info is given to VS as it arrives.) The result of a test suite is pretty much ignored, since it's not associated with any of the tests that have been discovered by VS. You can see that happen here: https://github.com/nunit/nunit3-vs-adapter/blob/master/src/NUnitTestAdapter/NUnitEventListener.cs#L148

The result of a test case is handled in the method right above that one. It records what is contained in the test case result.

---

@OsirisTerje commented on [Mon Dec 26 2016](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-269141549)

Thanks!  And what is the real meaning of the term/attribute "site" ?   It can take values like Child and Parent, but also SetUp and TearDown and the like ? 

---

@CharliePoole commented on [Mon Dec 26 2016](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-269142030)

In the framework, it's an enum. We use the enum names as attribute values. The default is "Test" which is what usually happens. The code in the adapter only logs a message for setup and teardown suite failures, since those don't get reflected anywhere else.

Note that the one time setup error message was being shown in the original issue description. That indicates it must be coming in the test case failure. You want to examine whether that's happening with the stack trace too. I suspect not.

---

@CharliePoole commented on [Mon Dec 26 2016](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-269142109)

BTW these attributes should all be documented here: https://github.com/nunit/docs/wiki/Test-Result-XML-Format

---

@carlin-q-scott commented on [Wed Apr 19 2017](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-295403429)

So if I want a OneTimeSetup that gives me sufficient detail to figure out the cause of the failure, should I do something like this?

```
static bool oneTimeSetupRan;

[SetUp]
public void OneTimeSetup()
{
    if (!oneTimeSetupRan)
    {
         throw new NotImplementedException();
         oneTimeSetupRan = true;
    }
}
```

---

@CharliePoole commented on [Wed Apr 19 2017](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-295461226)

@carlin-q-scott That would definitely simulate the one-time setup for a fixture. It's a pity to have to go to that trouble though.

---

@CharliePoole commented on [Wed Apr 19 2017](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-295462238)

@nunit/vs-extensions-team What do we want to do here? Is it simply a limitation of the VS interface? Should we try to report the stack trace for each error? That would require our remembering the error info for each one time setup failure and reporting it for the corresponding test cases. Non-trivial, especially in the presence of multiple fixtures running in parallel.

---

@blorger commented on [Fri Aug 18 2017](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-323297593)

This is not just limitation of VS interface. NUnit simply does not provide call stack information. Problem remains if you execute tests with NUnit console, no call stack on console output or in result file.
Why not simply change content of a message. Provide entire exception description rather than just exception message. This way there will be no need to debug a test or modify it (move one time setup code inside test, or wrap each setup method in try/catch block to log exception ...) just to get information where failure occurs.

---

@OsirisTerje commented on [Sun Oct 01 2017](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-333383025)

If I have an exception in the OneTimeSetup, for each test it is reported, like this:
OneTimeSetUp: ClassLibrary11.WhateverException : Exception of type 'ClassLibrary11.WhateverException' was thrown.

The stack trace is coming with the suite finished event a bit later, but then the test results have already been sent.  

I think it would be better just to return the full exception stacktrace with each failing tests.  It would only slow down failing tests, so that should be no big matter.  And then no changed code is needed for the adapter. 

@rprouse / @CharliePoole  If you agree, can we add an issue on NUnit for this, or move this to NUnit ? 

---

@CharliePoole commented on [Sun Oct 01 2017](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-333388234)

If we want to make it look a certain way for VS, then I think we should make it look that way in the adapter. NUnit itself attaches the error to the test that failed, which was the fixture. That's a pretty fundamental thing about how NUnit looks at tests: everything is a test - test cases, methods, classes, namespaces and the assembly. VS looks at the tests as a list of cases. Since the VS view is simpler, it should be possible to make things appear that way if it's really desired.

---

@OsirisTerje commented on [Sun Oct 01 2017](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-333400566)

Yes, we can do this in the adapter, but NUnit **is** reporting this per test case,  just with less information.  It has split the information of the exception between the test case and the fixture, with the test case reporting the exception and the message,  and the fixture reporting the same plus the stacktrace. 

---

@CharliePoole commented on [Mon Oct 02 2017](https://github.com/nunit/nunit3-vs-adapter/issues/284#issuecomment-333433563)

Add the issue to NUnit in that case, so we can discuss it there. I think it should be a new issue, since up to now this has been about how the adapter displays what it gets from NUnit, rather than changing what NUnit delivers.

We have a similar problem in the console runner, because the console runner historically only counted test cases and only displayed test case failures and errors. Over time, we started displaying more info. But the console has access to the entire XML, so modifying the output was easier.

**Comment by CharliePoole** [comment]
@OsirisTerje You moved the issue rather than adding one, which is OK, although I suspect we will end up having to do something in the adapter as well. However, the issue has no info in it about the framework, what it's doing wrong, how it should do things differently. How are you proposing the framework should be changed?

**Comment by rprouse** [comment]
With the new hierarchical test window in the latest version of Visual Studio, will this become a non-issue as the error will be reported at the fixture level?

**Comment by jnm2** [comment]
> With the new hierarchical test window in the latest version of Visual Studio, will this become a non-issue as the error will be reported at the fixture level?

I can confirm that it does not. The parent nodes are just group headers at present.

**Comment by jnm2** [comment]
So what should the adapter do?

For each test case, if the NUnit result site is Parent (for any kind of result), the adapter could go up the chain until the site is no longer Parent and convert that result to use as the test case's VSTest result. @OsirisTerje what do you think of that?

**Comment by tobyash86** [comment]
I am experiencing the same behavior. Is there any progress regarding this case?
Sometimes when fixture setup fails I need to debug it to find out where it has failed.

**Comment by jnm2** [comment]
There has been no progress on this beyond what you see.
If I understand this properly, either NUnit needs to start giving full stack trace information to child sites as well as the actual site, or else the adapter needs to compensate for the fact that it can't report parent nodes to VSTest and do the merge itself.

**Comment by CharliePoole** [comment]
Several of us suggested that this is not an issue for the framework at all, but the adapter. Should we perhaps decide and close this particular issue if we are not going to do anything on it?

**Comment by jnm2** [comment]
I do think that unless the adapter is blocked, we should close this one. I might not have the full picture so I wanted to wait for @rprouse or @OsirisTerje to confirm.

**Comment by jnm2** [comment]
⚠ @YuriyOborozhnyi brought this to my attention while working on https://github.com/nunit/nunit/issues/2906. This test run passes if you use the adapter:

```cs
public static class SomeFixture
{
    [Test]
    public static void SomeTest()
    {
    }

    [OneTimeTearDown]
    public static void SomeTearDown()
    {
        throw new System.Exception();
    }
}
```

This seems critically broken to me. NUnit 3.10.1, adapter 3.10.

Repeating https://github.com/nunit/nunit/issues/2466#issuecomment-366541760:

> There's no natural association from the once-per-fixture failures to any particular child test.

> Why don't we leave NUnit pure in this regard and do any shimming in the adapter like Charlie mentioned where necessary to translate the richer NUnit results to VSTest results?

@OsirisTerje My thinking right now is that if there is any fixture-level teardown, the adapter should not report any test results for tests in that fixture until the teardown happens. If there is an error in the teardown, all tests should error. What do you think? Can I open the issue back up in the adapter repo?

**Comment by CharliePoole** [comment]
Hmmm... The adapter doesn't have any way to know if there is fixture level teardown without duplicating the analysis done by the framework.

Of course, in one sense, there is always fixture-level teardown, but not always user-speciffied teardown. 

What about higher-level teardown? Let's say a teardown in a setup fixture fails. What would we do then? Taken to the extreme, we could say you have to wait till all nested suites are complete, right up to the assembly itself. But then that makes it a batch system, with no progress reported.

So, for all the above reasons, I think you have to do two things in the adapter:
1) Report exactly what nunit reports to you.
2) Additionally report any teardown errors.

**Comment by zmhh** [comment]
+1

**Comment by jamers99** [comment]
Any progress on this? This is very difficult to debug if there is a test failure in the set up or tear down (at any level). The stack trace is critical to debugging issues, especially in more complex scenarios.

**Comment by CharliePoole** [comment]
Since this is only a problem in the adapter, it seems odd that this issue remains in the framework project. If @OsirisTerje were to ask for something specific, then it would make sense. Otherwise, it's just an adapter issue.

**Comment by svengeance** [comment]
@CharliePoole I'm waiting for a response on the adapter on what the best way forward is - I've done a bit of work in it and I'd be comfortable making the change.

I'm concerned that the full exception isn't bubbled up from the Runner, though. I haven't fully delved into it, but it - might - need a change in more than one place.

**Comment by CharliePoole** [comment]
The framework definitely provides the stack trace, but possibly not in a way that's convenient for the adapter. That's why I thought @OsirisTerje might be the best person to request a framework change.

In my own use of the framework for the GUI runner, I'm able to handle this situation because the fixture result, containing the stack trace, is visible to the user. So I think there are two questions:

1. How may a runner, which doesn't normally show fixtures as separate entities, display such an error?
2. Should such a runner have to do the whole job or should the framework somehow assist?

**Comment by OsirisTerje** [comment]
Please check the enclosed adapter. It will output the stack trace for the exceptions in the overall message part. . It will be included in the trx file, and in Visual Studio Console Output window for testing.  This is the "easy" way of solving this.  It will not come out per test as the tests don't have that information from NUnit, and the work to retrieve that is not trivial.   But, with this change now, it will come out in the console output.  If needed, some more information can be included.  
[NUnit3TestAdapter.4.0.0-dev.751.zip](https://github.com/nunit/nunit/files/6564928/NUnit3TestAdapter.4.0.0-dev.751.zip)

**Comment by vladdex** [comment]
@OsirisTerje is there any resolution for this ? I've tried every solution but I cannot get Nunit to report exceptions that appear in OneTimeSetup methods.
Currently running NUnit3Test adapter 4.5.0

**Comment by OsirisTerje** [comment]
The stacktrace does flow up from the test:
![image](https://github.com/nunit/nunit/assets/203432/2c459420-a089-4653-8f19-9899e2631ad8)
Also from the command line:
![image](https://github.com/nunit/nunit/assets/203432/5d9322c5-9efc-4674-8283-21653c4ebdd7)

**Comment by manfred-brands** [comment]
@OsirisTerje The stack trace is only visible if going to the _Output_ tab and select `Tests` output. Not in the _Test Explorer_. There only the exception is shown. The stack trace doesn't seem to be propagated to the test result:

![image](https://github.com/nunit/nunit/assets/8655218/0447635e-dc92-40c2-9bdc-81c1237c0e60)

I tried to check this in the `OneTimeSetUpAndTearDownTests`, but the `StackTrace` is set in the fixure.
However the children have no `StackTrace`.
It gets lost in [CompositeWorkItem](https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Internal/Execution/CompositeWorkItem.cs#L96)

Is this intended or would you like for me to create a PR to pass the StackTrace to the `SkipChildren` and `SetChildWorkItemSkippedResult` methods?

**Comment by vladdex** [comment]
@OsirisTerje @manfred-brands 
I am seeing this issue when running the tests in an Azure Pipeline, sorry I forgot to mention that in my initial post.
This is how the output looks like
<img width="798" alt="image" src="https://github.com/nunit/nunit/assets/33092316/06774cf5-aa5a-4805-a698-2ea89a498aad">
<img width="723" alt="image" src="https://github.com/nunit/nunit/assets/33092316/056cd785-3870-4004-9c53-52fce8df567b">

**Comment by OsirisTerje** [comment]
@manfred-brands I am not sure what is intended or not here, but I can't see any reason to not propagate it further, so yes, please do that :-)  We can then get it out in the adapter to the tests. 

@vladdex true, the Azure Pipeline hides the console output messages.  I do believe you can get it out using the  `dotnet test --logger "console;verbosity=normal` (IIRC).  But if you use the built-in tasks i AP I don't think you can, but changing that in the yaml file should be easy (and if you use the classic pipeline setup, you should move over to yaml based). However, if @manfred-brands  can do the PR he suggests, we may get the stacktrace moved to the [first] tests.

**Comment by vladdex** [comment]
@OsirisTerje I tried your suggestion, I'm using the DotNetCoreCLI@2 task in the pipeline.
I get the stack in the console output of the tests, at the end of all test methods in the class, somewhere in the middle of the logs 
The stacktrace is still not present in the test results viewer from devops.
<img width="787" alt="image" src="https://github.com/nunit/nunit/assets/33092316/7a7dbf4e-ee99-428a-bbde-37427ce63188">
<img width="1120" alt="image" src="https://github.com/nunit/nunit/assets/33092316/fe04f518-8f97-440c-a7bf-cc816d6b13bf">

**Comment by OsirisTerje** [comment]
@vladdex That's good, as a workaround so far!  (I appreciate the screenshot, then we have doc for that situation too)

When @manfred-brands got the PR in, and a subsequent change in the adapter, we can get it out in the test results.  It will be a bit kind of fake, since the Test are actually not run at all, but from a users point of view it makes sense, so we'll do it. 
It will then be in 4.0 of NUnit and 4.6 of the adapter.

**Comment by manfred-brands** [comment]
I actually made the PR on v13.3. We were thinking of making a 13.4 before 4.0.
Will create one for 4.0 now.

**Comment by manfred-brands** [comment]
I have tested this with the Package artifact created from the build: https://github.com/nunit/nunit/actions/runs/6185437669

It now shows the stack trace at the test:
![image](https://github.com/nunit/nunit/assets/8655218/977cb5ef-92d3-4cd0-9320-2c60320b459b)
